File: algorithms/2_medium/R/01234_replace-the_substring_for_balanced_string/solution.c
static inline size_t get_idx(const char ch)
{
    switch (ch) {
        case 'Q':
            return 0;
        case 'W':
            return 1;
        case 'E':
            return 2;
        case 'R':
            return 3;
        default:
            fprintf(stderr, "unreachable!(): invalid input");
            exit(1);
    }
}

int balancedString(const char* s)
{
    const size_t len = strlen(s);
    assert(len % 4 == 0 && len != 0 && len <= 100000);
    int cnt[4] = { 0 };
    for (size_t i = 0; i < len; i += 1) {
        cnt[get_idx(s[i])] += 1;
    }
    size_t lo = 0;
    const int limit = (int)(len / 4);
    size_t res = len;

    for (size_t hi = 0; hi < len; hi +=1) {
        cnt[get_idx(s[hi])] -= 1;
        while (lo <= hi && cnt[get_idx(s[lo])] < limit) {
            cnt[get_idx(s[lo])] += 1;
            lo += 1;
        }

        if (cnt[0] <= limit && cnt[1] <= limit && cnt[2] <= limit && cnt[3] <= limit) {
            if (hi + 1 - lo < res) {
                res = hi + 1 - lo;
            }
        }
    }

    return (int)res;
}

File: algorithms/2_medium/R/01234_replace-the_substring_for_balanced_string/solution.rs
impl Solution {
    #[inline(always)]
    fn get_idx(b: u8) -> usize {
        match b {
            b'Q' => 0,
            b'W' => 1,
            b'E' => 2,
            b'R' => 3,
            _ => unreachable!()
        }
    }

    pub fn balanced_string(s: String) -> i32 {
        let bytes = s.as_bytes();
        let len = bytes.len();
        assert!(len % 4 == 0 && len != 0 && len <= 100_000);
        let mut cnt = [0_i32; 4];
        for &b in bytes {
            cnt[Self::get_idx(b)] += 1;
        }
        let mut lo = 0_usize;
        let limit = (len / 4) as i32;
        let mut res = len;

        for hi in 0..len {
            cnt[Self::get_idx(bytes[hi])] -= 1;
            while lo <= hi && cnt[Self::get_idx(bytes[lo])] < limit {
                 cnt[Self::get_idx(bytes[lo])] += 1;
                 lo += 1;
            }

            if cnt.iter().all(|c| *c <= limit) {
                 res = res.min(hi + 1 - lo);
            }
        }

        res.try_into().unwrap()
    }
}
